cxx/hirm.cc

```diff
@@ -16,6 +17,43 @@ HIRM::HIRM(const T_schema& schema, std::mt19937* prng) {
     }
   }
 }
+*/
+
+HIRM::HIRM(const T_schema& _schema, std::mt19937* prng) {
```

Comment:
This change is copied from https://github.com/probcomp/hierarchical-irm/pull/197.  The rationale is to prevent crashes when noisy relations have other noisy relations as their base relation.

Reply:
Where/why does it crash? I wouldn't have expected that.

Reply:
If you patch in this PR and then restore the commented out version of the HIRM constructor, the test test_make_pclean_samples will crash when constructing the HIRM.

The crash itself occurs in HIRM::add_relation inside the std::visit when schema.at(trel.base_relation) is called.

Reply:
Can you fix this by defining a new method that, when called on a noisy relation spec, recursively adds base relations if necessary? I think that would be more concise and easy to reason about, and as-is, it looks to me like the recursive call in `add_relation` will never execute. Alternatively, it seems fine to me to just initialize the `schema` member with the `schema` constructor arg instead of building it up in `add_relation`, which should also solve the crash. I just want to make sure we're not adding unnecessary complexity as we're about to hand it off to the MIT folks.

Reply:
1) Can we move this discussion to https://github.com/probcomp/hierarchical-irm/pull/197 which actually makes the change?

2) There are lots of ways you could fix this bug.  I don't think the recursive solution you propose is any more concise or easy to reason about than this one is, especially if we augment this solution with removing the then unneeded recursive call from add_relation.  But if you would like to code up any alternate solution, I would be happy to review it.